crates/matrix-sdk-common/src/linked_chunk/relational.rs

```diff
@@ -0,0 +1,658 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation for a _relational linked chunk_, see
+//! [`RelationalLinkedChunk`].
+
+use ruma::{OwnedRoomId, RoomId};
+
+use crate::linked_chunk::{ChunkIdentifier, Position, Update};
+
+/// A row of the [`RelationalLinkedChunk::chunks`].
+#[derive(Debug, PartialEq)]
+struct ChunkRow {
+    room_id: OwnedRoomId,
+    previous_chunk: Option<ChunkIdentifier>,
+    chunk: ChunkIdentifier,
+    next_chunk: Option<ChunkIdentifier>,
+}
+
+/// A row of the [`RelationalLinkedChunk::items`].
+#[derive(Debug, PartialEq)]
+struct ItemRow<Item, Gap> {
+    room_id: OwnedRoomId,
+    position: Position,
+    item: Either<Item, Gap>,
+}
+
+/// Kind of item.
+#[derive(Debug, PartialEq)]
+enum Either<Item, Gap> {
```

Comment:
Could we use a better name than `Either`? I thought it was a generic `Either<T, V>` enum, taken from the stdlib or an helper library.

Reply:
`ItemOrGap` :-D?